From 5b677644f756c685d9c465aa57efc4f51e456b8a Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Thu, 12 Dec 2013 16:51:55 +0000
Subject: [PATCH 2/3] Generalize CMakeLists.txt a bit ..

.. so that it works with 'MSYS Makefiles' and
'MinGW Makefiles' generators.

A warning was added about working around bison
needing BISON_SIMPLE env. var set at build time
---
 CMakeLists.txt | 28 ++++++++++++++++++++++------
 1 file changed, 22 insertions(+), 6 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4f10b15..dfffaa8 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -120,6 +120,21 @@ if (WIN32)
       hlslang/OSDependent/Windows/osinclude.h
       hlslang/OSDependent/Windows/ossource.cpp
     )
+
+    # The 'COMMAND set ..' doesn't work on MSYS or MinGW Makefiles because each
+    # command is executed in it's own shell.
+    IF ( MSYS )
+            message( WARNING "You will probably need to use:
+'BISON_SIMPLE=${CMAKE_CURRENT_SOURCE_DIR}/tools/bison.simple ${CMAKE_MAKE_PROGRAM}'
+.. to build this on MSYS")
+    ENDIF ( MSYS )
+
+    IF ( MINGW )
+            message( WARNING "You will probably need to use:
+'set \"BISON_SIMPLE=%CD%\\tools\\bison.simple\" & ${CMAKE_MAKE_PROGRAM}'
+.. to build this on MinGW")
+    ENDIF ( MINGW )
+
     source_group("OSDependent\\Windows" FILES ${OSDEPENDENT_FILES})
     include_directories(${CMAKE_CURRENT_SOURCE_DIR}/hlslang/OSDependent/Windows)
     
@@ -127,9 +142,8 @@ if (WIN32)
                          COMMAND set ARGS "BISON_SIMPLE=../../tools/bison.simple"
                          COMMAND set ARGS "BISON_HAIRY=../../tools/bison.simple"
                          COMMAND ../../tools/bison.exe ARGS  -d -t -v hlslang.y
-                         COMMAND copy ARGS /y hlslang_tab.c Gen_hlslang_tab.cpp
-                         COMMAND del ARGS hlslang_tab.c
-                         COMMAND del ARGS hlslang.output
+                         COMMAND ${CMAKE_COMMAND} -E copy_if_different hlslang_tab.c Gen_hlslang_tab.cpp
+                         COMMAND ${CMAKE_COMMAND} -E remove -f hlslang_tab.c hlslang.output
                          MAIN_DEPENDENCY hlslang/MachineIndependent/hlslang.y
                          WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/hlslang/MachineIndependent
                          COMMENT "Executing Bison on hlslang.y"
@@ -142,9 +156,11 @@ if (WIN32)
                          WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/hlslang/MachineIndependent
                          COMMENT "Executing flex on hlslang.l"
                       )
-    
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /D\"_HAS_ITERATOR_DEBUGGING=0\" /D\"_SECURE_SCL=0\" /D\"_CRT_SECURE_NO_WARNINGS\"")
-    SET(TEST_LIBS opengl32.lib)
+
+    IF ( MSVC )
+        SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /D\"_HAS_ITERATOR_DEBUGGING=0\" /D\"_SECURE_SCL=0\" /D\"_CRT_SECURE_NO_WARNINGS\"")
+    ENDIF ( MSVC )
+    SET(TEST_LIBS opengl32 gdi32 user32)
 elseif (APPLE)
     set(OSDEPENDENT_FILES
       hlslang/OSDependent/Mac/osinclude.h
-- 
1.8.4.4

